myia/abstract/infer.py

```diff
@@ -60,6 +64,60 @@
 )
 
 
+def _annotation_as_dict(type_hint):
+    """Return dict type args if type hint is a dict, else None."""
+    if type_hint is dict:
+        return ()
+    if (
+        isinstance(type_hint, typing._GenericAlias)
+        and type_hint.__origin__ is dict
+    ):
+        return type_hint.__args__
+    return None
+
+
+def _check_dict_annotation(abstract, type_args):
+    """Check if abstract is an abstract dict.
+
+    type_args is either () or the couple (key_type, value_type) of
+    expected python types for dict keys and values.
+    """
+    if not isinstance(abstract, AbstractDict):
+        raise TypeMismatchError(AbstractDict, abstract)
+    if type_args:
+        key_type, value_type = type_args
+        if key_type is not str:
+            raise TypeMismatchError(str, key_type)
+        abstract_value_type = type_to_abstract(value_type)
+        for child in abstract.children():
+            annotation_merge(
+                abstract_value_type, child, forced=True, bind_pending=True
+            )
+    return abstract
+
+
+def validate_annotation(annotation, abstract):
+    """Check if abstract is allowed by given annotation."""
+    try:
+        dict_annotation = _annotation_as_dict(annotation)
+        if dict_annotation is not None:
+            # type_to_abstract could not generate a suitable
+            # AbstractDict from a type hint because it won't
+            # provide dict keys and values, so we need
+            # a specific check for dict types.
+            _check_dict_annotation(abstract, dict_annotation)
```

Comment:
Could `type_to_abstract` simply be extended to fix this issue?

Reply:
Hi @breuleux ! I implemented `pytype_to_abstract` to support `dict` type hints. Annotation tests pass, however, abstract dict need to know keys in some other places in myia code. So, I added a kind of `placeholder` class to set AbstractDict entries when keys are unknown, and I updated `amerge(AbstractDict, AbstractDict)` to support the placeholder. Could you check it ? (last commit)